[Monitoring] Metricbeat migration net new user experience

chrisronline opened the pull request:
This PR is step 2, part 2 of the [Metricbeat setup wizard](https://github.com/elastic/kibana/issues/39010) and is also known as the "net new user" experience.

The main purpose of this PR is to provide a UX for new users to enable monitoring for their stack products. Currently, this PR is only concerned with Elasticsearch and Kibana, but a future PR will add support for other stack products (This is because there is/was outstanding work to enable metricbeat collection for these other products).

The current UX for enabling monitoring is extremely simple - it's just a button click in the UI. However, because of the need to install and start a beat, the process is slightly more complicated but the work done in this PR should help the user understand how to enable monitoring, and also guide the user through the experience too.

## Testing

### Elasticsearch monitoring
- Start up ES
- Start up Kibana
- Load Kibana in the browser
- Navigate to the stack monitoring application
- You should land on the `no-data` page which is the default experience until step 4 is merged
- To overwrite this behavior to test this PR, manually navigate to http://localhost:5601/app/monitoring#/loading?_g=(inSetupMode:!t) which will route you through the new user setup flow created through this PR
- Follow instructions on the page to enable Elasticsearch monitoring with Metricbeat

### Kibana monitoring
- Follow up steps above to enable Elasticsearch monitoring with Metricbeat
- Navigate to the stack monitoring application, ensuring setupMode is enabled. 
  - The url should look like: http://localhost:5601/app/monitoring#/overview?_g=(cluster_uuid:yhZt4d92TLKQd8YtH-JO-Q,inSetupMode:!t)
- Scroll down to see the Kibana section on the overview page
- Note that you'll see `Instances: 0 (0/1)`
- Click `Instances: 0`
- Follow instructions on the page to enable Kibana monitoring through Metricbeat

### Edge cases

- Please test with multiple instances of ES
- I don't really see a need to test multiple instances of Kibana, as there is no way for us to _know_ ahead of time that there is another kibana instance (so we can suggest they add monitoring to it).

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5086/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5111/)

elasticmachine commented:
Pinging @elastic/stack-monitoring

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5329/)

ycombinator commented:
> Start up ES - ensure any monitoring settings are disabled
> Start up Kibana, ensuring the xpack.monitoring.kibana.collection.enabled flag is set to false

I'm not sure these steps should be a prerequisite for testing the UX introduced by this PR. A net new user will start with default settings. So we should make sure the UX introduced by this PR works with these default settings, right?

ycombinator commented:
> Load Kibana in the browser
> Navigate to the stack monitoring application
> Note that you will no longer see the "No data screen", instead you will be taken to the Elasticsearch nodes listing page

Does this mean that when this PR is merged the user will start seeing the new UX already? If so, I thought we don't want this to be the case until the Step 4 PR is merged?

chrisronline commented:
> Does this mean that when this PR is merged the user will start seeing the new UX already? If so, I thought we don't want this to be the case until the Step 4 PR is merged, right?

Yes, this needs to some more work before it is reviewable. Please hold off on reviews for now. I'll re-add reviewers once this is sorted out. Thanks

chrisronline commented:
@ycombinator @igoristic This is ready for review now. I've updated the PR description to reflect the testing steps now required based on the changes suggested by @ycombinator

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5633/)

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/metricbeat_migration/flyout/flyout.js:
```diff
@@ -22,17 +22,19 @@ import {
 } from '@elastic/eui';
 import { getInstructionSteps } from '../instruction_steps';
-import { Storage } from '../../../../../../../../src/legacy/ui/public/storage/storage';
+import { Storage } from '../../../../../../../../src/legacy/ui/public/storage';
```
You can also do short hand paths here eg: `import { Storage } from 'ui/storage';`

igoristic commented on x-pack/legacy/plugins/monitoring/public/directives/main/index.js:
```diff
@@ -118,6 +119,19 @@ export class MonitoringMainController {
     return this._licenseService.mlIsSupported();
   }
+
+  isDisabledTab(product, _tab) {
```
Looks like we don't need `_tab` here, and we can also remove this parameter from views when calling the function

igoristic commented on x-pack/legacy/plugins/monitoring/public/lib/setup_mode.js:
```diff
@@ -30,6 +30,15 @@ const setupModeState = {
+  const executor = angularState.injector.get('$executor');
+  globalState.cluster_uuid = clusterUuid;
+  globalState.save();
+  angularState.scope.$apply();
```
Maybe worth wrapping it in a safe handler instead? eg:
```javascript
angularState.scope.$apply(() => (globalState.cluster_uuid = clusterUuid));
```

chrisronline replied:
Would that work? We'd still need to call `.save()` though? What are the benefits of this approach?

igoristic replied:
It's more optimal in that it'll only digest scope on what ever expression is passed, and I'll wrap it in try/catch incase there is already a digest in-progress: https://docs.angularjs.org/api/ng/type/$rootScope.Scope#$apply

You can put `globalState.save();` in the $apply handler as well, if you think it has any bindings to update

igoristic commented on x-pack/legacy/plugins/monitoring/public/views/loading/index.js:
```diff
@@ -0,0 +1,69 @@
+            return toggleSetupMode(true)
+              .then(() => {
+                kbnUrl.changePath('/elasticsearch/nodes');
+                $scope.$apply();
```
I think `kbnUrl.changePath()` will already do a digest, so probably no need for `$scope.$apply()`

chrisronline replied:
If that's the case, would I see an error about already in the digest loop?

igoristic replied:
It might be done one after the other, but it does not really matter since `kbnUrl.changePath` will completely tear down that component/page and replace it with the new components. This means that everything will already be updated and synced

igoristic commented on x-pack/legacy/plugins/monitoring/public/views/loading/index.js:
```diff
@@ -0,0 +1,69 @@
+        const kbnUrl = $injector.get('kbnUrl');
+
+        setAngularState($scope, $injector);
+        initSetupModeState();
```
Ref **A**: I know this is more from the previous PR, but is there any reason why we need to set state first and then use it? Maybe some kind of static singleton (where we don't keep state) approach would work better here eg:
```javascript
initSetupModeState($scope, $injector);
```
Do that for all exported handlers in `setup_mode.js` that depend on 'em, and remove `setAngularState` all together.

Might be nitpicking though

chrisronline replied:
Great suggestion! I'll change it

igoristic commented on x-pack/legacy/plugins/monitoring/public/views/loading/index.js:
```diff
@@ -0,0 +1,69 @@
+              return;
+            }
+            setAngularState($scope, $injector);
+            return toggleSetupMode(true)
```
See ref **A**

igoristic commented on x-pack/legacy/plugins/monitoring/server/lib/setup/collection/get_collection_status.js:
```diff
@@ -285,6 +311,25 @@ export const getCollectionStatus = async (req, indexPatterns, clusterUuid) => {
+
+    if (product.name === KIBANA_SYSTEM_ID && liveKibanaInstance) {
+      productStatus.byUuid = {
+        [get(liveKibanaInstance, 'kibana.uuid')]: {
```
Are you using a `get()` here incase it does not exist? Wouldn't the key be `undefined` then?

chrisronline replied:
It's definitely not guaranteed to be there so I'm just trying to be defensive

igoristic replied:
Right, so if it's not guaranteed to be there you might end up with something like:
```javascript
[undefined]: {...
```
Might be same issue for: https://github.com/elastic/kibana/pull/39832/files/6a9a348d871795b272c683b5cc2bbdc1040591f1#diff-a3c36eaa7da8d988333f17352cfce93cR154

Maybe add a backup key? eg: `[get(liveKibanaInstance, 'kibana.uuid') || 'uuid']: {...`

Or, maybe not even add that property all together if the `uuid` is not there?

Just an observation and some nitpicking. I might be totally wrong here (I often am)

igoristic commented:
Looks pretty good overall 💪 

Just some minor feedback/suggestions

ycombinator commented:
The node name in the Node Listing table is clickable while in Setup Mode. Clicking it causes an error.

![Jul-09-2019 04-10-03](https://user-images.githubusercontent.com/51061/60870879-a9434000-a1ff-11e9-993e-306b26500de9.gif)

Perhaps the node name shouldn't be a link while in Setup Mode?

ycombinator commented:
A net new user, by definition, doesn't have any monitoring setup. So telling them to "migrate" doesn't make sense IMO.

<img width="1320" alt="Screen Shot 2019-07-09 at 4 12 22 AM" src="https://user-images.githubusercontent.com/51061/60871016-f0c9cc00-a1ff-11e9-840d-f2c44971a150.png">

ycombinator commented:
In the same vein as my previous comment, "migration" doesn't make sense for a net new user.

<img width="564" alt="Screen Shot 2019-07-09 at 4 20 29 AM" src="https://user-images.githubusercontent.com/51061/60871581-0b507500-a201-11e9-8110-95273f8ec33c.png">

ycombinator commented:
I really like the `1/1` and `0/1` indicators.

<img width="296" alt="Screen Shot 2019-07-09 at 4 23 07 AM" src="https://user-images.githubusercontent.com/51061/60871741-55d1f180-a201-11e9-8a19-0bbfc2b812be.png">

But if I was a user, especially a net new one, I'm not sure it would be obvious to me what these stand for. Perhaps we should add a tooltip?

ycombinator commented:
Similar to my [earlier comment](https://github.com/elastic/kibana/pull/39832#issuecomment-509538986), the instance name in the Kibana Instance Listing table is a link, even when Setup Mode is enabled. Clicking on it generates an error.

![Jul-09-2019 04-26-28](https://user-images.githubusercontent.com/51061/60872013-ed374480-a201-11e9-888a-4718da3e57b6.gif)

Perhaps we should disable this link in Setup Mode?

ycombinator commented:
I started up a 1-node ES cluster and then setup monitoring for that node per the instructions in this PR. Then I started up a 2nd ES node in the same cluster. After waiting a few seconds (while still in Setup Mode), the Cluster Overview page looks like this:

<img width="287" alt="Screen Shot 2019-07-09 at 4 34 21 AM" src="https://user-images.githubusercontent.com/51061/60872669-1d331780-a203-11e9-9bde-2d4534cfaeb5.png">

Notice how Nodes shows `2` but the little box on the top right shows `1/1`. I would expect it to start showing `1/2` as soon as Nodes updates to 2 as well.

If it helps, after manually refreshing the page, the little box on the top right changes to `1/2`.

chrisronline commented:
@ycombinator @igoristic I've addressed your feedback and this is ready for another round of review.

@ycombinator In terms of [this comment](https://github.com/elastic/kibana/pull/39832#issuecomment-509543548), I've decided to go with this visual, using EUI colors to indicate status. LMK what you think
<img width="485" alt="Screen Shot 2019-07-10 at 9 49 12 AM" src="https://user-images.githubusercontent.com/56682/60981865-1f7d9a80-a305-11e9-91a9-8fb909f97fb3.png">

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6631/)

ycombinator commented:
I took another pass at this PR. All my previous feedback has been addressed.

> I've decided to go with this visual, using EUI colors to indicate status.

I like this change. I also noticed the tooltips and found them to be helpful. I think the wording in them could be improved but that's for later as part of overall language cleanup anyway.

I noticed https://github.com/elastic/kibana/pull/39832#issuecomment-509548218 hasn't been addressed yet but chatting off-PR, I know you were thinking of addressing this in a follow up PR also across the board. Can you add this to the meta issue so we don't lose track of it?

Finally, one use case that I hadn't previously thought of:

What if a net new user has Beats or Logstash or APM Server also running and wants to know how to setup Metricbeat monitoring for those. Should we show sections for these on the cluster overview page with buttons linking to setup instructions (or something else to the same effect)?

chrisronline commented:
> Can you add this to the meta issue so we don't lose track of it?

Added.

> What if a net new user has Beats or Logstash or APM Server also running and wants to know how to setup Metricbeat monitoring for those. Should we show sections for these on the cluster overview page with buttons linking to setup instructions (or something else to the same effect)?

Agreed. This will be added/covered in Step 3 from the meta

chrisronline commented:
@igoristic I think I addressed all your points, but let me know if I missed something and I'd be happy to fix it up!

igoristic commented:
> @igoristic I think I addressed all your points, but let me know if I missed something and I'd be happy to fix it up!

@chrisronline Everything looks good for the most part 👍 Also, looks great functionally :smile:

Still a little concerned about:
https://github.com/elastic/kibana/pull/39832#discussion_r302709725
https://github.com/elastic/kibana/pull/39832#discussion_r302703846

But, might not be a big deal here, and probably just a matter of preference

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8272/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8294/)

chrisronline commented:
@ycombinator @igoristic Some changes came in to support our existing api tests that you might want to review here: https://github.com/elastic/kibana/pull/39832/commits/c63327c49fcc97077a3f47d27269338195893a6e

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8455/)

spalger commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8479/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8658/)

chrisronline commented:
Backport:

7.x: https://github.com/elastic/kibana/commit/01d4b698b7f2d729227fd95190a7ffd1a41a4af0
